File: src/picross.js
// Import Pako for binary compression
import * as Pako from "../lib/pako.js";
window.Pako = Pako;
//Puzzle also has other methods below this definition
export class Puzzle {
    //Constructors
    constructor(dimension = 0, size = []) {
        //Number of dimensions of the puzzle
        this.dimension = dimension;
        //Array with the length in each dimension
        this.size = size.slice();
        this.calculateSizes();
        //List of blocks
        this.shape = new Array(this.shapeSize);
        //List of hints, which is the maximum face size multiplied by the dimension
        this.hintsTotal = new Array(this.maxFaceSize * dimension);
        this.hintsPieces = new Array(this.maxFaceSize * dimension);
        //Metadata like name, color...
        this.metadata = {};
    }
    calculateSizes() {
        //Computer total size of shape
        this.shapeSize = 1;
        let minimumRowLength = 10000;
        for(let i = 0; i < this.dimension; i++) {
            this.shapeSize *= this.size[i];
            minimumRowLength = Math.min(minimumRowLength, this.size[i]);
        }
        //Max face size is the total size divided by the smallest row.
        this.maxFaceSize = this.shapeSize / minimumRowLength;
        //Spacing is a cumulative size metric. If this.size=[4,3,2], this.spacing=[1,4,12]
        this.spacing = [];
        let spacing = 1;
        for(let i = 0; i < this.dimension + 1; i++) {
            this.spacing.push(spacing);
            spacing *= this.size[i];
        }
    }
}
//Positional retrieval functions
//Turns position vector into an index
Puzzle.prototype.collapsePos = function (vector) {
    //Multiply each position component by its spacing and take the sum
    return vector.map((p, i) => p * this.spacing[i]).reduce((a, b) => a + b);
}
//Splices a row from a starting position index and a direction
Puzzle.prototype.getRow = function (pos, dim) {
    let out = [];
    let spacing = this.spacing[dim];
    //Start is the position at the beginning of the row
    let start = pos % spacing + Math.floor(pos / this.spacing[dim + 1]) * this.spacing[dim + 1];
    for(let i = 0; i < this.size[dim]; i++) {
        out[i] = this.shape[start + spacing * i];
    }
    return out;
}
//Convert a position and direction into a hint position in the hint array
Puzzle.prototype.getHintPosition = function (position, direction) {
    let below = position % this.spacing[direction];
    let above = Math.floor(position / this.spacing[direction + 1]) * this.spacing[direction];
    return this.maxFaceSize * direction + below + above;
}
//Converts multidimensional position index into 3d position index where dims is the slicer state
Puzzle.prototype.positionTo3D = function (vector, dims) {
    let size = [];
    for(let i in [0, 1, 2]) size[i] = this.size[dims.indexOf(-1 - i)] || 1;
    let posArray = dims.slice();
    //Set x,y, and z of the output array
    posArray[dims.indexOf(-1)] = vector % size[0];
    posArray[dims.indexOf(-2)] = Math.floor(vector / size[0]) % size[1];
    posArray[dims.indexOf(-3)] = Math.floor(vector / size[0] / size[1]) % size[2];
    //Collapse to 3d position
    return this.collapsePos(posArray);
}
//Converts position index into position vector
Puzzle.prototype.getVector = function (position) {
    let out = [];
    for(let i = 0; i < this.dimension; i++) {
        out[i] = Math.floor(position / this.spacing[i]) % this.size[i];
    }
    return out;
}
//Extra data generation
//Generate all hints for the entire puzzle (puzzle should not have indetermined pieces)
Puzzle.prototype.generateHints = function () {
    //Reset hint arrays
    this.hintsPieces = new Array(this.maxFaceSize * this.dimension);
    this.hintsPieces.fill(0, 0, this.hintsPieces.length);
    this.hintsTotal = new Array(this.maxFaceSize * this.dimension);
    this.hintsTotal.fill(0, 0, this.hintsTotal.length);
    //Iterate through each hint
    this.foreachHint((cell, t, p, dim, i) => {
        let row = this.getRow(cell, dim);
        //Count total and how many pieces its in
        let total = 0, pieces = 0, prev = cell_broken;
        for(let x = 0; x < row.length; x++) {
            if(row[x] == cell_colored) {
                total++;
                if(prev == cell_broken) pieces++;
            }
            prev = row[x];
        }
        if(total == 0) pieces = 1;
        this.hintsPieces[i + this.maxFaceSize * dim] = pieces;
        this.hintsTotal[i + this.maxFaceSize * dim] = total;
    });
}
//Generate visible sides of a 3d puzzle
Puzzle.prototype.generateSidesVisible = function () {
    this.visibleSides = [];
    let spacing = [1, this.size[0], this.size[0] * this.size[1]];
    this.foreachCell((cell, i, pos) => {
        if(cell == cell_broken) return this.visibleSides[i] = 0;
        let visible = 0;
        for(let dim = 0; dim < this.dimension; dim++) {
            let indexInRow = Math.floor(i / this.spacing[dim]) % this.size[dim];
            let spa = spacing[dim];
            if(indexInRow == 0 || this.shape[i - spa] == cell_broken) visible |= (4 ** dim);
            if(indexInRow == this.size[dim] - 1 || this.shape[i + spa] == cell_broken) visible |= (4 ** dim) * 2;
        }
        this.visibleSides[i] = visible;
    });
}
//Create a new puzzle from a puzzle and a slicer state
Puzzle.prototype.project3D = function (dims, puz) {
    //Find x, y, and z axis
    this.size = [];
    for(let i in [0, 1, 2]) this.size[i] = puz.size[dims.indexOf(-1 - i)] || 1;
    this.dimension = 3;
    this.calculateSizes();
    this.shape = new Array(this.shapeSize);
    this.hintsTotal = new Array(this.maxFaceSize * this.dimension);
    this.hintsPieces = new Array(this.maxFaceSize * this.dimension).fill(0);
    //Copy cells
    this.foreachCell((cell, i, pos) => {
        //Find position in original map
        let oldPos = dims.slice();
        for(let i in [0, 1, 2]) oldPos[dims.indexOf(-1 - i)] = pos[i];
        this.shape[i] = puz.shape[puz.collapsePos(oldPos)];
    });
    //Copy hints
    this.foreachHint((cell, total, pieces, dim, i) => {
        let hintPos = i + dim * this.maxFaceSize;
        if(dims.indexOf(-1 - dim) == -1) {
            this.hintsTotal[hintPos] = 0;
            this.hintsPieces[hintPos] = 0;
            return;
        }
        let position = [];
        for(let i = 0; i < this.dimension; i++) {
            position[i] = cell % this.size[i];
            cell = Math.floor(cell / this.size[i]);
        }
        position[dim] = 0;
        let newPosition = dims.slice();
        for(let i = 0; i < 3; i++) newPosition[dims.indexOf(-1 - i)] = position[i];
        let oldHintPosition = puz.getHintPosition(puz.collapsePos(newPosition), dims.indexOf(-1 - dim));
        this.hintsTotal[hintPos] = puz.hintsTotal[oldHintPosition];
        this.hintsPieces[hintPos] = puz.hintsPieces[oldHintPosition];
    });
}
//Intersect start and velocity vector with the first solid cube
Puzzle.prototype.rayIntersect = function (start, vel) {
    let velocityMagnitude = Math.sqrt(vel[0] * vel[0] + vel[1] * vel[1] + vel[2] * vel[2]);
    //Setup starting variables
    let normVel = vel.map(v => v / velocityMagnitude);
    let dir = vel.map(v => Math.sign(v));
    let closestCell = start.map((v, i) => ((v < 0 ? 1 - v : v) % 1 + 1) % 1);
    let tDelta = normVel.map(v => Math.abs(1 / v));
    let tMax = tDelta.map((v, i) => closestCell[i] * v);
    //Variables that keep track of state
    let curCell = start.map(v => Math.floor(v));
    let recentFace = 0;
    let pos = this.collapsePos(curCell);
    //Raycast loop
    for(let count = 0; count < 100; count++) {
        //Check if intersects cube
        if(curCell.every(v => v >= 0) && curCell.every((v, i) => v < this.size[i])) {
            if(scene.voxels[pos].visible) return {pos: pos, face: (recentFace * 2) + (dir[recentFace] < 0 ? 1 : 0)};
        }
        //Advance along closest plane to tip of ray
        let closest = Math.min(tMax[0], tMax[1], tMax[2]);
        for(let i = 0; i < 3; i++) if(closest == tMax[i]) {
            tMax[i] += tDelta[i];
            curCell[i] += dir[i];
            recentFace = i;
            pos += dir[i] * this.spacing[i];
            break;
        }
    }
    return {pos: -1, face: -1};
}
//Generate a new puzzle with certain hints missing based on a linear difficulty scale
Puzzle.prototype.fromDifficulty = function (difficulty) {
    //Random number generator seeded with a
    function mulberry32(a) {
        return function () {
            var t = a += 0x6D2B79F5;
            t = Math.imul(t ^ t >>> 15, t | 1);
            t ^= t + Math.imul(t ^ t >>> 7, t | 61);
            return ((t ^ t >>> 14) >>> 0) / 4294967296;
        }
    }
    let rand = mulberry32(0xDEADBEEF);
    let out = Puzzle.fromString(this.toString());
    out.foreachHint((cell, total, pieces, dim, i) => {
        if(rand() < difficulty) out.hintsPieces[dim * out.maxFaceSize + i] = 0;
    });
    return out;
}
//Estimate time to solve puzzle
Puzzle.prototype.estimateTime = function () {
    Module.setPuzzle(this.toString());
    return Module.solve();
}
//Foreach functions
Puzzle.prototype.foreachCell = function (func) {
    let pos = new Array(this.dimension);
    pos.fill(0);
    for(let i = 0; i < this.shapeSize; i++) {
        func(this.shape[i], i, pos);
        pos[0]++;
        let x = -1;
        while(pos[++x] >= this.size[x]) {
            pos[x] = 0;
            pos[x + 1]++;
        }
    }
}
//Iterate over each hint and run func(cell,total,pieces,dimension,index)
Puzzle.prototype.foreachHint = function (func) {
    for(let dim = 0; dim < this.dimension; dim++) {
        for(let i = 0; i < this.shapeSize / this.size[dim]; i++) {
            let cell = Math.floor(i / this.spacing[dim]) * this.spacing[dim + 1] + i % this.spacing[dim];
            let hintPos = i + dim * this.maxFaceSize;
            func(cell, this.hintsTotal[hintPos], this.hintsPieces[hintPos], dim, i);
        }
    }
}
//Utility Functions
//Delete all rows that have a zero for a hint
Puzzle.prototype.deleteZeroedRows = function () {
    this.foreachHint((cell, total, pieces, dim) => {
        if(total == 0 && pieces == 1) {
            for(let i = 0; i < this.size[dim]; i++) {
                this.shape[cell + this.spacing[dim] * i] = cell_broken;
            }
        }
    });
}
Puzzle.prototype.generateActionableRows = function () {

}
//Export functions
Puzzle.prototype.toString = function () {
    const A = 'A'.charCodeAt(0);
    //Sizes and dimension
    let out = this.dimension + "~"
        + this.size.map(size => String.fromCharCode(A + size)).join("") + "~"
        //Shape
        + this.shape.map(cell => cell == cell_unsure ? "-" : (cell == cell_colored ? "+" : " ")).join("") + "~";
    //Hints
    for(let i = 0; i < this.dimension; i++) {
        for(let x = 0; x < this.shapeSize / this.size[i]; x++) {
            let index = i * this.maxFaceSize + x;
            out += String.fromCharCode(A + this.hintsTotal[index], A + this.hintsPieces[index]);
        }
    }
    out += "~" + JSON.stringify(this.metadata);
    return out;
}
//Not sure what this is supposed to be for
Puzzle.prototype.toUintArray = function () {
    let meta = JSON.stringify(this.metadata);
    let length = 16;
    let hintLength = this.size.reduce((p, v) => p + this.shapeSize / v, 0);
    let sizes = [this.dimension + 1, hintLength, Math.ceil(this.shapeSize / 4), meta.length];
    console.log(sizes);
    for(let i = 0; i < 4; i++) length += sizes[i];
    let out = new Uint8ClampedArray(length);
    for(let i = 0; i < 4; i++) {
        let size = sizes[i];
        for(let x = 3; x >= 0; x--) {
            out[i * 4 + x] = size % 256;
            size = Math.floor(size / 256);
        }
    }
    let pos = 17;
    out[16] = this.dimension;
    for(let i = 0; i < this.dimension; i++) {
        out[pos++] = this.size[i];
    }
    console.log(pos);
    for(let i = 0; i < this.dimension; i++) {
        for(let x = 0; x < this.shapeSize / this.size[i]; x++) {
            let index = i * this.maxFaceSize + x;
            out[pos++] = this.hintsTotal[index] << 3 + this.hintsPieces[index];
        }
    }
    console.log(pos);
    pos--;
    for(let i = 0; i < this.shapeSize; i++) {
        if(i % 4 == 0) pos++;
        out[pos] |= this.shape[i] << (2 * i);
    }
    console.log(pos);
    pos++;
    for(let i = 0; i < meta.length; i++) {
        out[pos++] = meta.charCodeAt(i);
    }
    console.log(pos);
    return out;
}
//Convert puzzle to a base64 string that can be shared
Puzzle.prototype.toBase64 = function () {
    let string = this.toString();
    return btoa(String.fromCharCode.apply(null, Pako.deflate(string)));
}
//Import functions
//Create puzzle from raw string
Puzzle.fromString = function (str) {
    const A = 'A'.charCodeAt(0);
    let strs = str.split("~");
    let size = strs[1].split("").map(c => c.charCodeAt(0) - A);
    let puz = new Puzzle(Number(strs[0]), size, "");
    for(let i = 0; i < strs[2].length; i++) {
        let char = strs[2].charAt(i);
        if(char == '+') puz.shape[i] = cell_colored;
        else if(char == ' ') puz.shape[i] = cell_broken;
        else if(char == '-') puz.shape[i] = cell_unsure;
        else puz.shape[i] = cell_error;
    }
    puz.hintsTotal = new Array(puz.maxFaceSize * puz.dimension);
    puz.hintsPieces = new Array(puz.maxFaceSize * puz.dimension);
    let index = 0;
    for(let dim = 0; dim < puz.dimension; dim++) {
        for(let x = 0; x < puz.shapeSize / puz.size[dim]; x++) {
            let hintPos = dim * puz.maxFaceSize + x;
            puz.hintsTotal[hintPos] = strs[3].charCodeAt(index) - A;
            puz.hintsPieces[hintPos] = strs[3].charCodeAt(index + 1) - A;
            index += 2;
        }
    }
    puz.metadata = JSON.parse(strs[4] || "{}");
    return puz;
}
//Create puzzle from base64
Puzzle.fromBase64 = function (str) {
    try {
        let unencoded = atob(str);
        let data = new Uint8Array(unencoded.length);
        unencoded.split("").map((a, i) => data[i] = a.charCodeAt(0));
        let unzippedString = Pako.inflate(data, {to: "string"});
        return Puzzle.fromString(unzippedString);
    }
    catch(e) {
        printError("Invalid puzzle");
    }
}
const cell_error = 0;
const cell_broken = 1;
const cell_colored = 2;
const cell_unsure = 3;

File: src/picross.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <cstring>
#include <vector>
#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#include <emscripten/bind.h>
#endif

using namespace std;
typedef enum Cell : char {
    unsure = 3, broken = 1, painted = 2
} Cell;
typedef struct Hint {
    unsigned char total;
    unsigned char pieceCount;
} Hint;
void solveRow(int total, int pieces, int rowLength, Cell* row, Cell* out);
void printCells(Cell* cells, int count) {
    const char* type = "0 +-EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE";
    for(int i = 0;i < count;i++) cout << type[cells[i]];
}
class Puzzle {
public:
    string name;
    int dimension;
    int* size = NULL;
    Cell* shape = NULL;
    Hint* hints = NULL;
private:
    int maxFaceSize;
    int shapeSize;
    int* spacing = NULL;
    void computeShapeSize() {
        int minimumRowLength = 10000;
        shapeSize = 1;
        for(int i = 0;i < dimension;i++) {
            shapeSize *= size[i];
            if(size[i] < minimumRowLength) minimumRowLength = size[i];
        }
        maxFaceSize = shapeSize / minimumRowLength;
        free(spacing);
        spacing = (int*)calloc(dimension, sizeof(int));
        int space = 1;
        for(int i = 0;i < dimension;i++) {
            spacing[i] = space;
            space *= size[i];
        }
    }
public:
    Puzzle(string str) {
        string strs[10];
        int currentIndex = 0;
        int i = 0;
        //Split by tildes
        int lastTilde = -1;
        int x = 0;
        for(int i = 0;i < str.size() + 1;i++) {
            if(str[i] == '~' || str[i] == 0) {
                strs[x] = str.substr(lastTilde + 1, i - lastTilde - 1);
                lastTilde = i;
                x++;
            }
        }
        //Set name
        name = strs[0];
        //Dimension and size
        dimension = atoi(strs[1].c_str());
        free(size);
        size = (int*)calloc(dimension, sizeof(int));
        for(int i = 0;i < dimension;i++) size[i] = strs[2][i] - 'A';
        computeShapeSize();
        //Copy shape
        free(shape);
        shape = (Cell*)calloc(strs[3].size(), sizeof(Cell));
        memset(shape, 3, strs[3].size());
        for(int i = 0;i < strs[3].size();i++) {
            if(strs[3][i] == ' ') shape[i] = broken;
            else if(strs[3][i] == '-') shape[i] = unsure;
            else if(strs[3][i] == '+') shape[i] = painted;
        }
        //Copy hints
        free(hints);
        hints = (Hint*)calloc(dimension * maxFaceSize, sizeof(Hint));
        int index = 0;
        for(int dim = 0; dim < dimension; dim++) {
            for(int x = 0; x < shapeSize / size[dim]; x++) {
                int hintPos = dim * maxFaceSize + x;
                hints[hintPos].total = strs[4][index] - 'A';
                hints[hintPos].pieceCount = strs[4][index + 1] - 'A';
                index += 2;
            }
        }
    }
    Puzzle(int dim, int* siz, string nam) {
        dimension = dim;
        free(size);
        size = (int*)calloc(dim, sizeof(int));
        memcpy(size, siz, dim * sizeof(int));
        name = nam;
        computeShapeSize();
        free(shape);
        shape = (Cell*)calloc(shapeSize, sizeof(Cell));
    }
    void solve() {
        bool isChange = true;
        int changeCount = 0;
        int loopCount = 0;
        while(isChange) {
            isChange = false;
            for(int dim = 0;dim < dimension;dim++) {
                Cell row[size[dim]];
                Cell out[size[dim]];
                for(int i = 0;i < shapeSize;i++) {
                    memset(out, 0, size[dim]);
                    //If not first cell in row
                    if((i / spacing[dim]) % size[dim] != 0) continue;
                    Hint hint = hints[getRowPosition(i, dim) + dim * maxFaceSize];
                    getRow(i, dim, row);
                    solveRow(hint.total, hint.pieceCount, size[dim], row, out);
                    if(memcmp(row, out, size[dim] * sizeof(Cell)) != 0) {
                        changeCount++;
                        isChange = true;
                        setRow(i, dim, out);
                    }
                }
            }
            loopCount++;
            cout << "Change count " << loopCount << ": " << changeCount << endl;
        }
        cout << "Change count: " << changeCount << endl;
        cout << "Loop count: " << loopCount << endl;
    }
    void generateHints() {
        free(hints);
        hints = (Hint*)calloc(maxFaceSize * dimension, sizeof(Hint));
        for(int i = 0;i < shapeSize;i++) {
            for(int dim = 0;dim < dimension;dim++) {
                int rowPos = getRowPosition(i, dim);
                //To prevent duplicate calculations
                if(hints[rowPos + dim * maxFaceSize].pieceCount != 0) continue;
                int spacing = 1;
                for(int x = 0;x < dimension;x++) spacing *= size[x];
                //Set i to first cell in the row
                i = ((i / spacing) / size[dim]) * spacing * size[dim] + i % spacing;
                int total = 0, pieces = 0, prev = broken;
                for(int x = 0;x < size[dim];x++) {
                    int cell = i + x * spacing;
                    if(shape[cell] == painted) {
                        total++;
                        if(prev == broken) pieces++;
                    }
                    prev = shape[cell];
                }
                hints[rowPos + dim * maxFaceSize].pieceCount = pieces;
                hints[rowPos + dim * maxFaceSize].total = total;
            }
        }
    }
    int getRowPosition(int pos, int direction) {
        //Example:
        // size = {5,2,3,4}
        // pos = 48, direction = 1
        // below = 8
        // above = 10
        int below = pos % spacing[direction];
        int above = (pos / spacing[direction] / size[direction]) * spacing[direction];
        return below + above;
    }
    void posToPositionArray(int pos, int* out) {
        for(int i = 0;i < dimension;i++) {
            out[i] = pos % size[i];
            pos /= size[i];
        }
    }
    string toString() {
        // NAME~DIMENSION~SIZE~SHAPE~HINTS
        string out = name + "~" + std::to_string(dimension) + "~";
        for(int i = 0;i < dimension;i++) {
            out += string(1, 'A' + size[i]);
        }
        //Shape
        out += "~";
        for(int i = 0;i < shapeSize;i++) {
            if(shape[i] == broken) out += " ";
            else if(shape[i] == unsure) out += "-";
            else if(shape[i] == painted) out += "+";
        }
        //Hints
        out += "~";
        for(int dim = 0;dim < dimension;dim++) {
            for(int x = 0;x < shapeSize / size[dim];x++) {
                int ind = maxFaceSize * dim + x;
                out += string(1, char('A' + hints[ind].total)) + string(1, char('A' + hints[ind].pieceCount));
            }
        }
        return out;
    }
    int collapsePosition(int* position) {
        int out = 0;
        for(int i = 0;i < dimension;i++) {
            out += position[i] * spacing[i];
        }
        return out;
    }
    Cell getCell(int* position) {
        return shape[collapsePosition(position)];
    }
    Hint getHint(int* position, int direction) {
        int pos = 0;
        for(int i = 0;i < dimension;i++) {
            if(i == direction) continue;
            pos += position[i] * (spacing[i] / spacing[direction]);
        }
        return hints[maxFaceSize * direction + pos];
    }
    void setCell(int* position, Cell set) {
        shape[collapsePosition(position)] = set;
    }
    void getRow(int position, int dim, Cell* out) {
        int rowLength = size[dim];
        position = position % spacing[dim] + ((position / spacing[dim] / size[dim])) * spacing[dim] * size[dim];
        for(int i = 0;i < rowLength;i++) {
            out[i] = shape[position + spacing[dim] * i];
        }
    }
    void setRow(int position, int dim, Cell* cells) {
        int rowLength = size[dim];
        position = position % spacing[dim] + ((position / spacing[dim] / size[dim])) * spacing[dim] * size[dim];
        for(int i = 0;i < rowLength;i++) {
            shape[position + spacing[dim] * i] = cells[i];
        }
    }
};
Puzzle* puz;
typedef struct SpacingPossiblity {
    unsigned char space[8];
    void print() {
        for(int i = 0;i < 8;i++) cout << int(space[i]) << " ";
        cout << endl;
    }
} SpacingPossibility;
SpacingPossibility NULLPOSSIB = { {0,0,0,0,0,0,0,0} };

vector<SpacingPossibility> spacePossibilities(int total, int pieces, bool zeroStart, bool zeroEnd) {
    vector<SpacingPossibility> out;
    if(pieces == 1) {
        out.push_back(NULLPOSSIB);
        out[0].space[0] = total;
        return out;
    }
    else if(pieces == 2) {
        int x = 0;
        for(int i = zeroStart ? 0 : 1;i <= total - (zeroEnd ? 0 : 1);i++) {
            SpacingPossibility possib;
            possib.space[0] = i;
            possib.space[1] = total - i;
            out.push_back(possib);
            x++;
        }
        return out;
    }
    else {
        int margin = total - pieces + (zeroEnd ? 1 : 0) + 1;
        for(int i = zeroStart ? 0 : 1;i <= margin;i++) {
            vector<SpacingPossibility> possib = spacePossibilities(total - i, pieces - 1, false, zeroEnd);
            int size = out.size();
            for(int x = 0;x < possib.size();x++) {
                out.push_back(NULLPOSSIB);
                out[size + x].space[0] = i;
                memcpy(out[size + x].space + 1, possib[x].space, 7);
            }
        }
    }
    return out;
}
void orCells(Cell* dest, Cell* in, int len) {
    for(int i = 0;i < len;i++) dest[i] = Cell(int(in[i]) | int(dest[i]));
}
bool possibilityMatchesRow(Cell* row, Cell* possibility, int len) {
    for(int i = 0;i < len;i++) {
        if((int(row[i]) | int(possibility[i])) != int(row[i])) return false;
    }
    return true;
}
void solveRow(int total, int pieces, int rowLength, Cell* row, Cell* out) {
    if(pieces == 0) {
        memcpy(out, row, rowLength);
        return;
    }
    if(total == 0) {
        memset(out, 0, rowLength);
        return;
    }
    //Generate possibilities for broken run
    vector<SpacingPossibility> bkn = spacePossibilities(rowLength - total, pieces + 1, true, true);
    //Generate possibilities for colored runs
    vector<SpacingPossibility> crd = spacePossibilities(total, pieces, false, false);
    //Go through all possibilities
    for(int b = 0;b < bkn.size();b++) {
        //Check if either all unsure or matches current row to break
        if(bkn[b].space[0] != 0 && row[0] == painted) continue;
        for(int c = 0;c < crd.size();c++) {
            Cell possib[rowLength];
            memset(possib, 0, rowLength);
            int pos = 0;
            int countTotal = bkn[b].space[0];
            memset(possib, broken, countTotal);
            for(int j = 0;j < pieces;j++) {
                int count = crd[c].space[j];
                memset(possib + countTotal, painted, count);
                countTotal += count;
                int gap = bkn[b].space[j + 1];
                memset(possib + countTotal, broken, gap);
                countTotal += gap;
            }
            if(possibilityMatchesRow(row, possib, rowLength)) {
                orCells(out, possib, rowLength);
            }
        }
    }
}
void stringToRow(string str, Cell* row, int len) {
    memset(row, 3, len);
    for(int i = 0;i < str.size();i++) {
        if(str[i] == ' ') row[i] = broken;
        if(str[i] == '+') row[i] = painted;
    }
};
string rowToString(Cell* row, int len) {
    string out(len, ' ');
    char types[4] = { 'E', ' ', '+', '?' };
    for(int i = 0;i < len;i++) {
        out[i] = types[int(row[i])];
    }
    out[len] = 0;
    return out;
}
string solveRowJS(int total, int pieces, int rowLength, string row) {
    Cell rowCells[rowLength];
    Cell out[rowLength];
    stringToRow(row, rowCells, rowLength);
    memset(out, 0, rowLength);
    solveRow(total, pieces, rowLength, rowCells, out);
    return rowToString(out, rowLength);
}
string getPuzzle() {
    return puz->toString();
}
void setPuzzle(string str) {
    puz = new Puzzle(str);
}
void doHints() {
    puz->generateHints();
}
void solveCurrentPuzzle() {
    puz->solve();
}
#ifdef __EMSCRIPTEN__
EMSCRIPTEN_BINDINGS(a) {
    emscripten::function("solveRow", &solveRowJS);
    emscripten::function("getPuzzle", &getPuzzle);
    emscripten::function("setPuzzle", &setPuzzle);
    emscripten::function("generateHints", &doHints);
    emscripten::function("solve", &solveCurrentPuzzle);
}
int main() {
    return 0;
}
#else
int main() {
    puz = new Puzzle("Basic puzzle~3~MHF~------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------~AAAAAAAAAABBBBCCCCGBGBCBCBBBAAAAGBJBFCFCBBCCCCGBGBCBCBBBAAAAAAAAAABBBBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACBAAAAAAAAAAAAAAAAAAAACBHBCBCBCBCBEBAAAAAAAABBCBEBCBCBCBCBCBBBCCEBCBAACBHBCBCBCBCBEBAAAAAAAAAAAACBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACCAAAAAAAACCAAAAAAAAAAAACCAAAAAAAACCAAAAAAAAAAAADBDBDBDBDBDBAAAAAAAAAAAADBDBDBDBDBDBBBBBBBAABBDBDBAAAAAAAAAAAAAABBBBAADBFBAAAAAAAAAAAABBBBBBAAAAECAAAAAAAAAAAAAABBAA");
    puz->solve();
    cout << puz->toString();
}
#endif
